### ISSUE TITLE
pr_auc give unintuitive value for constant prediction

### ISSUE THREAD

**Issue by klin333** [tidymodels/yardstick#229]
When logistic regression model predicts 1 constant value (eg a model with an intercept only ie predict the sample probability always), the pr_auc function does not give intuitive value. It is likely due to the use of trapezoidal rule for calculating the area under curve. Maybe have to ignore the area under the curve to the left of the smallest predicted value? 

I don't know if it is 'wrong' but at least currently it is not intuitive. For example, a glmnet might give unintuitively high pr_auc as the L1 penalty goes really high to the point of 1 intercept term only, due to this problem with pr_auc. 

The pr_curve() autoplot() also seems non intuitive due to the linear interpolation making the pr curve look a lot better than it should be. 

```
library(tidyverse)

N <- 10000
p <- 0.01
set.seed(100)

df <- tibble(truth_factor = factor(sample(c('TRUE', 'FALSE'), N, replace = TRUE, prob = c(p, 1-p)), levels = c('TRUE', 'FALSE'))) %>%
  mutate(truth_num = if_else(truth_factor == 'TRUE', 1, 0)) %>%
  mutate(pred_const = 0.01) %>%
  mutate(pred_const2 = if_else(row_number() < n()/2, 0, 1)) %>%
  mutate(pred_rand = runif(n()))

yardstick::pr_auc(df, truth = truth_factor, pred_rand)$.estimate     # 0.02377628 ok
MLmetrics::PRAUC(df$pred_rand, df$truth_num)                         # 0.01515559
yardstick::pr_curve(df, truth = truth_factor, pred_rand) %>% autoplot()

yardstick::pr_auc(df, truth = truth_factor, pred_const)$.estimate    # 0.5058 wrong
MLmetrics::PRAUC(df$pred_const, df$truth_num)                        # 0
yardstick::pr_curve(df, truth = truth_factor, pred_const) %>% autoplot()

yardstick::pr_auc(df, truth = truth_factor, pred_const2)$.estimate    # 0.2194559 still wrong
MLmetrics::PRAUC(df$pred_const2, df$truth_num)                        # 0.006179606
yardstick::pr_curve(df, truth = truth_factor, pred_const2) %>% autoplot()


```

**Comment by DavisVaughan** [comment]
The TLDR here is that you should use `average_precision()` rather than `pr_auc()`, as it avoids these pathological edge cases.

---

This has come up before. The difference between yardstick and MLmetrics has to do with the value of `precision` when `recall = 0` and you don't have any false positives yet. We use `precision = 1`, while MLmetrics uses `precision = NaN`. There are arguments for both, and there is no consensus among which is best. We used to use `precision = NA`, but switched to `precision = 1` for various reasons in this PR https://github.com/tidymodels/yardstick/pull/95. Switching back to `NA` would just cause others to argue that it should have been `1`.

It is worth noting that sklearn also uses `precision = 1` in this edge case.

```r
library(reticulate)
library(yardstick)
library(purrr)
library(dplyr)

skmetrics <- import("sklearn.metrics")

N <- 10000
p <- 0.01
set.seed(100)

df <- tibble(truth_factor = factor(sample(c('TRUE', 'FALSE'), N, replace = TRUE, prob = c(p, 1-p)), levels = c('TRUE', 'FALSE'))) %>%
  mutate(truth_num = if_else(truth_factor == 'TRUE', 1, 0)) %>%
  mutate(pred_const = 0.01) %>%
  mutate(pred_const2 = if_else(row_number() < n()/2, 0, 1)) %>%
  mutate(pred_rand = runif(n()))

mlmetrics_pr_curve <- function(truth, estimate) {
  pred_obj <- ROCR::prediction(estimate, truth)
  perf_obj <- ROCR::performance(pred_obj, measure = "prec", x.measure = "rec")
  tibble(recall = perf_obj@x.values[[1]], precision = perf_obj@y.values[[1]])
}

pr_curve(df, truth_factor, pred_const)
#> # A tibble: 2 x 3
#>   .threshold recall precision
#>        <dbl>  <dbl>     <dbl>
#> 1     Inf         0    1     
#> 2       0.01      1    0.0116
mlmetrics_pr_curve(df$truth_factor, df$pred_const)
#> # A tibble: 2 x 2
#>   recall precision
#>    <dbl>     <dbl>
#> 1      0  NaN     
#> 2      1    0.0116

res <- skmetrics$precision_recall_curve(df$truth_factor, df$pred_const, pos_label = "TRUE")
skmetrics$auc(res[[2]], res[[1]])
#> [1] 0.5058

pr_auc(df, truth_factor, pred_const)
#> # A tibble: 1 x 3
#>   .metric .estimator .estimate
#>   <chr>   <chr>          <dbl>
#> 1 pr_auc  binary         0.506
```

Instead, I'd encourage you to look into the `average_precision()` metric. I added it specifically because it avoids this edge case (by design), see https://github.com/tidymodels/yardstick/pull/95#issuecomment-473455580.

```r
average_precision(df, truth_factor, pred_rand)
#> # A tibble: 1 x 3
#>   .metric           .estimator .estimate
#>   <chr>             <chr>          <dbl>
#> 1 average_precision binary        0.0245
average_precision(df, truth_factor, pred_const)
#> # A tibble: 1 x 3
#>   .metric           .estimator .estimate
#>   <chr>             <chr>          <dbl>
#> 1 average_precision binary        0.0116
average_precision(df, truth_factor, pred_const2)
#> # A tibble: 1 x 3
#>   .metric           .estimator .estimate
#>   <chr>             <chr>          <dbl>
#> 1 average_precision binary        0.0108
```

It is also listed in the sklearn docs as being a nice alternative to PR AUC:
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html#sklearn.metrics.average_precision_score

> This implementation is not interpolated and is different from computing the area under the precision-recall curve with the trapezoidal rule, which uses linear interpolation and can be too optimistic.
